components/centraldashboard/app/stackdriver_metrics_service_test.ts

```diff
@@ -0,0 +1,378 @@
+import * as monitoring from '@google-cloud/monitoring';
+
+import {Interval} from './metrics_service';
+import {StackdriverMetricsService} from './stackdriver_metrics_service';
+
+describe('StackdriverMetricsService', () => {
+  const projectId = 'test-project';
+  const nodeNames = ['node-1', 'node-2'];
+  const emptyTimeSeriesPromise =
+      Promise.resolve([[]] as [gapi.client.monitoring.TimeSeries[]]);
+  const aggregation = {
+    alignmentPeriod: {
+      seconds: 60,
+    },
+    perSeriesAligner: 'ALIGN_MEAN',
+    crossSeriesReducer: 'REDUCE_NONE',
+  };
+  let stackdriverClient: jasmine.SpyObj<monitoring.MetricServiceClient>;
+  let service: StackdriverMetricsService;
+
+  beforeEach(() => {
+    jasmine.clock().install();
+    jasmine.clock().mockDate(new Date(1557705600000));
+    stackdriverClient = jasmine.createSpyObj<monitoring.MetricServiceClient>(
+        'mockStackdriverClient', ['listTimeSeries', 'projectPath']);
+    stackdriverClient.projectPath.withArgs(projectId).and.returnValue(
+        `projects/${projectId}`);
+    service =
+        new StackdriverMetricsService(stackdriverClient, projectId, nodeNames);
+  });
+
+  afterEach(() => {
+    jasmine.clock().uninstall();
+  });
+
+  describe('Node CPU Utilization', () => {
+    it('Retrieves for multiple nodes', async () => {
+      const response =
+          ([
+            {
+              metric: {labels: {instance_name: 'node-1-gce-instance-abc'}},
+              points: [
+                {
+                  interval: {
+                    startTime: {seconds: 1557705600},
+                    endTime: {seconds: 1557705600},
+                  },
+                  value: {
+                    doubleValue: 0.08,
+                  },
+                },
+                {
+                  interval: {
+                    startTime: {seconds: 1557705300},
+                    endTime: {seconds: 1557705300}
+                  },
+                  value: {
+                    doubleValue: 0.1,
+                  },
+                },
+              ]
+            },
+            {
+              metric: {labels: {instance_name: 'node-2-gce-instance-xyz'}},
+              points: [
+                {
+                  interval: {
+                    startTime: {seconds: 1557705600},
+                    endTime: {seconds: 1557705600},
+                  },
+                  value: {
+                    doubleValue: 0.07,
+                  },
+                },
+                {
+                  interval: {
+                    startTime: {seconds: 1557705300},
+                    endTime: {seconds: 1557705300},
+                  },
+                  value: {
+                    doubleValue: 0.12,
+                  },
+                },
+              ]
+            }
+          ] as unknown) as gapi.client.monitoring.TimeSeries[];
```

Comment:
Why cast it as unknown first?

Reply:
Since I am not building a mock with all of the indicated type's properties, Typescript's compiler complains that the object given does not sufficiently overlap with the specified type. By casting it to unknown first, you're able to work around the compiler's warning and say "I know what I am doing in this case". It would generally not be a good idea to do that in production code but for mocking test responses where we only use a subset of the properties in the object, it's a reasonable pattern.

Reply:
Gotcha, thanks for the explanation!